stdlib/Random/src/Xoshiro.jl

```diff
@@ -0,0 +1,231 @@
+    t = s1 << 17
+    s2 = xor(s2, s0)
+    s3 = xor(s3, s1)
+    s1 = xor(s1, s2)
+    s0 = xor(s0, s3)
+    s2 = xor(s2, t)
+    s3 = s3 << 45 | s3 >> 19
+    task.rngState0, task.rngState1, task.rngState2, task.rngState3 = s0, s1, s2, s3
+    res
+end
+
+# Shared implementation between Xoshiro and TaskLocalRNG -- seeding
+function seed!(rng::Union{TaskLocalRNG, Xoshiro}, seed::UInt128)
+    seed0 = seed % UInt64
+    seed1 = (seed>>>64) % UInt64
+    seed!(rng, (seed0, seed1, zero(UInt64), zero(UInt64)))
+end
+seed!(rng::Union{TaskLocalRNG, Xoshiro}, seed::Integer) = seed!(rng, UInt128(seed))
+
+seed!(rng::Union{TaskLocalRNG, Xoshiro}) =
+    seed!(rng, (rand(RandomDevice(), UInt64), rand(RandomDevice(), UInt64),
+                rand(RandomDevice(), UInt64), rand(RandomDevice(), UInt64)))
+
+function seed!(rng::Union{TaskLocalRNG, Xoshiro}, seed::AbstractVector{UInt64})
+    if length(seed) > 4
+        throw(ArgumentError("seed should have no more than 256 bits"))
+    end
+    seed0 = length(seed)>0 ? seed[1] : UInt64(0)
+    seed1 = length(seed)>1 ? seed[2] : UInt64(0)
+    seed2 = length(seed)>2 ? seed[3] : UInt64(0)
+    seed3 = length(seed)>3 ? seed[4] : UInt64(0)
+    seed!(rng, (seed0, seed1, seed2, seed3))
+end
+
+function seed!(rng::Union{TaskLocalRNG, Xoshiro}, seed::AbstractVector{UInt32})
+    if iseven(length(seed))
+        seed!(rng, reinterpret(UInt64, seed))
+    else
+        seed!(rng, UInt64[reinterpret(UInt64, @view(seed[begin:end-1])); seed[end] % UInt64])
+    end
+end
+
+@inline function rand(rng::Union{TaskLocalRNG, Xoshiro}, ::SamplerType{UInt128})
+    first = rand(rng, UInt64)
+    second = rand(rng,UInt64)
+    second + UInt128(first)<<64
+end
+
+@inline rand(rng::Union{TaskLocalRNG, Xoshiro}, ::SamplerType{Int128}) = rand(rng, UInt128) % Int128
+
+@inline function rand(rng::Union{TaskLocalRNG, Xoshiro},
+                      T::SamplerUnion(Bool, Int8, UInt8, Int16, UInt16, Int32, UInt32, Int64))
+    S = T[]
+    # use upper bits
+    (rand(rng, UInt64) >>> (64 - 8*sizeof(S))) % S
+end
+
+function copy(rng::TaskLocalRNG)
+    t = current_task()
+    Xoshiro(t.rngState0, t.rngState1, t.rngState2, t.rngState3)
+end
+
+function copy!(dst::TaskLocalRNG, src::Xoshiro)
+    t = current_task()
+    t.rngState0, t.rngState1, t.rngState2, t.rngState3 = src.s0, src.s1, src.s2, src.s3
+    dst
+end
+
+function copy!(dst::Xoshiro, src::TaskLocalRNG)
+    t = current_task()
+    dst.s0, dst.s1, dst.s2, dst.s3 = t.rngState0, t.rngState1, t.rngState2, t.rngState3
+    dst
+end
+
+function ==(a::Xoshiro, b::TaskLocalRNG)
+    t = current_task()
+    a.s0 == t.rngState0 && a.s1 == t.rngState1 && a.s2 == t.rngState2 && a.s3 == t.rngState3
+end
+
+==(a::TaskLocalRNG, b::Xoshiro) = b == a
+
+# for partial words, use upper bits from Xoshiro
+
+rand(r::Union{TaskLocalRNG, Xoshiro}, ::SamplerTrivial{UInt52{UInt64}}) = rand(r, UInt64) >>> 12
+
+function rand(r::Union{TaskLocalRNG, Xoshiro}, ::SamplerTrivial{UInt104{UInt128}})
```

Comment:
This must return a `UInt128` with the low 104 bits random, and the upper ones 0.
But there is a generic function for this, better to define the lower level sampler for `::SamplerTrivial{UInt104Raw{UInt128}}`, which is the same except it doesn't care the the upper bits are 0. One possible implementation would be `rand(r, UInt128)`, or `rand(r, UInt128) >> 24` if you want to use the upper bits (then it matches the non-"raw" version.
Actually, the default implementation for `UInt104Raw` would be good, as it uses two `UInt52Raw`. So I suggest to replace the two `rand` methods above by:
```julia
rand(r::Union{TaskLocalRNG, Xoshiro}, ::SamplerTrivial{UInt52Raw{UInt64}}) = rand(r, UInt64) >>> 12
# both methods below optional, to avoid having to mask in the generic method:
rand(r::Union{TaskLocalRNG, Xoshiro}, ::SamplerTrivial{UInt52{UInt64}}) = rand(r, UInt64) >>> 12
rand(r::Union{TaskLocalRNG, Xoshiro}, ::SamplerTrivial{UInt104{UInt128}}) = rand(r, UInt104Raw())
```